fastserde/avro-fastserde-tests-common/src/test/java/com/linkedin/avro/fastserde/logical/types/LogicalTypesTestBase.java

```diff
@@ -0,0 +1,220 @@
+                    new TimeConversions.TimestampMicrosConversion());
+
+            for (Conversion<?> conversion : missingConversions) {
+                modelData.addLogicalTypeConversion(conversion);
+            }
+        }
+    }
+
+    protected <T> byte[] serialize(FastSerializer<T> fastSerializer, T data) throws IOException {
+        InMemoryEncoder encoder = new InMemoryEncoder();
+        fastSerializer.serialize(data, encoder);
+        return encoder.toByteArray();
+    }
+
+    protected <T> byte[] serialize(DatumWriter<T> datumWriter, T data) throws IOException {
+        InMemoryEncoder encoder = new InMemoryEncoder();
+        datumWriter.write(data, encoder);
+
+        return encoder.toByteArray();
+    }
+
+    protected Instant toInstant(Object maybeDate) {
+        if (maybeDate == null) {
+            return null;
+        } else if (maybeDate instanceof LocalDate) {
+            return ((LocalDate) maybeDate).atStartOfDay(ZoneId.systemDefault()).toInstant();
+        } else if (maybeDate instanceof LocalDateTime) {
+            return ((LocalDateTime) maybeDate).toInstant(ZoneOffset.UTC);
+        } else {
+            throw new UnsupportedOperationException(maybeDate + " is not supported (yet)");
+        }
+    }
+
+    /**
+     * Runs various serializers and ensures each one returned the same result.
+     * @return result from default serialization, normally it's just {@code data.toByteBuffer().array()}
+     */
+    protected <T extends SpecificRecordBase> byte[] verifySerializers(T data,
+            FunctionThrowingIOException<T, ByteBuffer> toByteBuffer) throws IOException {
+        // given
+        Schema schema = data.getSchema();
+        FastSerializer<T> fastGenericSerializer = new FastGenericSerializerGenerator<T>(schema, classesDir, classLoader, null)
+                .generateSerializer();
+
+        FastSerializer<T> fastSpecificSerializer = new FastSpecificSerializerGenerator<T>(schema, classesDir, classLoader, null)
+                .generateSerializer();
+
+        GenericDatumWriter<T> genericDatumWriter = new GenericDatumWriter<>(
+                schema, copyConversions(data.getSpecificData(), new GenericData()));
+
+        SpecificDatumWriter<T> specificDatumWriter = new SpecificDatumWriter<>(
+                schema, copyConversions(data.getSpecificData(), new SpecificData()));
+
+        fixConversionsIfAvro19(data.getSpecificData());
+
+        // when
+        byte[] fastGenericBytes = serialize(fastGenericSerializer, data);
+        byte[] fastSpecificBytes = serialize(fastSpecificSerializer, data);
+        byte[] genericBytes = serialize(genericDatumWriter, data);
+        byte[] specificBytes = serialize(specificDatumWriter, data);
+        byte[] bytesWithHeader = toByteBuffer.apply(data).array(); // contains 10 extra bytes at the beginning
+
+        byte[] expectedHeaderBytes = ByteBuffer.wrap(new byte[v1HeaderLength])
```

Comment:
All of this play with header bytes is due to the fact that our custom InMemoryEncoder doesn't add them like the default BinaryEncoder does? Maybe we should move the header related code to the InMemoryEncoder?

Reply:
I tried to move the V1 header to `InMemoryEncoder` locally and after all I think such move complicates serializing, e.g. `serialize()` helper method would need `schema` parameter:
```
serialize(fastGenericSerializer, data);
```
which might be confusing.

Reply:
In `serialize()` methods we need to change `<T>` to `<T extends GenericRecord>` which seems to be fine.

Added in commit `[review-fixes] Part 6: InMemoryEncoder adds V1 header.` (will be pushed soon)